### PR TITLE
Add fullscreen and app crop selection options

### PR DESCRIPTION
**Pull request by karaggeorge**
This PR adds the following features: https://github.com/wulkano/kap/issues/300, https://github.com/wulkano/kap/issues/298, https://github.com/wulkano/kap/issues/212

In order to get this to work, I made the following package: https://github.com/karaggeorge/mac-windows that returns the windows in the current screen along with their information and given an application name it brings it to the front

I wasn't sure what the design should look like. For now, I just added another select next to the aspect ratio one. 

When an Application is selected, if a user either resizes the crop window or changes the width/height inputs the select will unselect the application. Selecting an Application will also set the appropriate width/height/aspect-ratio in the other inputs.

The package I made only allows the main window of an Application to come to the front. For example, if you have 3 different Finder windows open, only the last used one (main) will show up. The AppKit Swift library only supports that for now.

screen shot 2017-12-12 at 16 03 27

Gifs
--
### Recording an App:
kapture 2017-12-12 at 13 26 27

### Recording Fullscreen:
kapture 2017-12-12 at 13 29 31

### Changing dimensions unselects Application
kapture 2017-12-12 at 13 30 41

### Refreshing open applications
kapture 2017-12-12 at 15 50 46

And the Gifs created by the first two:
--
kapture 2017-12-12 at 13 25 50

kapture 2017-12-12 at 13 28 55

### CODE REVIEW

**Comment by doot0** [comment]
This is _a m a z i n g_ work, @karaggeorge, especially as a first-time contributor. Targeting specific windows has been a want of ours for a long and we've not had the bandwidth to implement, so THANK YOU! <3

Provided the team's happy with changes then I'm absolutely down for merging this.

**Comment by karaggeorge** [comment]
I would hold up on the merge for a little bit. I'm still trying to fix some small issues (like when recording full screen you can still see the cropper window). If people could download the branch and test it that would be great. It would help catch minor bugs before the merge

**Comment by albinekb** [comment]
I think it should refresh the list on click, show a spinner if it takes more than 500ms to do so and then trannsition to a menu? Really nice!!

**Comment by karaggeorge** [comment]
@albinekb At the moment, it refreshes everytime the application is shown. It replaces the `Select...` choice with `Loading...` while the list is being loaded. I just thought that on-click would be too often, and you can't really open/close a new application without blurring from Kap (which would cause it to hide and reload the list again on show). Thoughts?

File: app/src/renderer/js/main.js [hunk 1]
```diff
@@ -406,6 +485,8 @@ document.addEventListener('DOMContentLoaded', () => {
 
   ipcRenderer.on('log', (event, msgs) => console.log(...msgs));
 
+  ipcRenderer.on('load-applications', loadApplications);
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `load-applications` => `load-apps`

File: app/src/renderer/js/main.js [hunk 2]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+    appSelector.querySelector('option[disabled]').selected = true;
+  }
+
+  appSelector.append(createOption('Fullscreen'));
+  appSelector.append(createOption('──────────', true));
+  appSelector.addEventListener('change', handleAppChange);
+
+  function loadApplications() {
+    // Remove existing applications
+    appSelector.querySelectorAll('option:not([disabled])').forEach(option => {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: Use a `for-of` loop.

File: app/src/renderer/js/main.js [hunk 3]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  function clearApp() {
+    appSelector.querySelector('option[disabled]').selected = true;
+  }
+
+  appSelector.append(createOption('Fullscreen'));
+  appSelector.append(createOption('──────────', true));
+  appSelector.addEventListener('change', handleAppChange);
+
+  function loadApplications() {
+    // Remove existing applications
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `applications` => `apps`

File: app/src/renderer/js/main.js [hunk 4]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+    appSelector.querySelectorAll('option:not([disabled])').forEach(option => {
+      if (option.value !== 'Fullscreen') {
+        option.remove();
+        delete appData[option.value];
+      }
+    });
+
+    appSelector.querySelector('option[disabled]').text = 'Loading...';
+    appSelector.querySelector('option[disabled]').selected = true;
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: You're using `appSelector.querySelector('option[disabled]')` 4 times. Save it to a variable.

File: app/src/renderer/js/main.js [hunk 5]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+
+    // Load applications
+    getWindows().then(windows => {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: Use `await`

File: app/src/renderer/js/main.js [hunk 6]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      appSelector.querySelector('option[disabled]').text = 'Select...';
+      windows.forEach(window => {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: Use a `for-of` loop.

File: app/src/renderer/js/main.js [hunk 7]
```diff
@@ -370,9 +440,18 @@ document.addEventListener('DOMContentLoaded', () => {
     recordBtn.attributes['data-state'].value = 'initial';
   });
 
+  ipcRenderer.on('cropper-window-opened', (event, bounds) => {
+    recordBtn.classList.add('is-cropping');
+    recordBtn.attributes['data-state'].value = 'ready-to-record';
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: This should use `recordBtn.dataset`.

File: app/src/renderer/views/main.pug [hunk 8]
```diff
@@ -29,6 +29,10 @@ block body
             option(value='3:2') 3:2
             option(value='1:1' selected) 1:1
             option(id="custom-ratio-option" data-aspect-ratio='custom') Custom
+        +select-custom.has-icon.c-select--large.app-select
+          +icon('ic_crop_free_24px').o-icon--small
+          select.app-selector(tabindex="1")
+            option(disabled=true selected=true) Loading...
```
Comment (sindresorhus) [inline app/src/renderer/views/main.pug]: `Loading...` => `Loading…`

File: app/src/renderer/js/main.js [hunk 9]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  function loadApplications() {
+    // Remove existing applications
+    appSelector.querySelectorAll('option:not([disabled])').forEach(option => {
+      if (option.value !== 'Fullscreen') {
+        option.remove();
+        delete appData[option.value];
+      }
+    });
+
+    appSelector.querySelector('option[disabled]').text = 'Loading...';
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `Loading...` => `Loading…`

File: app/src/renderer/js/main.js [hunk 10]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+    appSelector.querySelector('option[disabled]').selected = true;
+
+    // Load applications
+    getWindows().then(windows => {
+      appSelector.querySelector('option[disabled]').text = 'Select...';
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `Select...` => `Select…`

File: app/src/renderer/js/main.js [hunk 11]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
 
+  const createOption = (label, disabled = false) => {
+    const option = document.createElement('option');
+    option.value = label;
+    option.text = label;
+    option.disabled = disabled;
+    return option;
+  };
+
+  const appData = {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: This should be a `Map`, not object.

File: app/src/renderer/js/main.js [hunk 12]
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
 
   const [micOnIcon, micOffIcon] = toggleAudioRecordBtn.children;
 
+  const createOption = (label, disabled = false) => {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `disabled` here creates a [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) when used. Use an options object instead.

File: app/src/renderer/js/main.js [hunk 13]
```diff
@@ -14,6 +15,8 @@ const {app} = remote;
 // Observers that should be disposed when the window unloads
 const observersToDispose = [];
 
+const cropperWindowBuffer = 2;
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: It's not clear enough from the name what this is and why it's 2. Needs a succinct code comment.

File: app/src/main/index.js [hunk 14]
```diff
@@ -173,6 +177,25 @@ ipcMain.on('open-cropper-window', (event, size) => {
...
     });
   }
+};
+
+ipcMain.on('activate-application', (event, appName, {width, height, x, y}) => {
+  if (cropperWindow) {
+    cropperWindow.close();
+  }
+
+  activateWindow(appName).then(() => {
```
Comment (sindresorhus) [inline app/src/main/index.js]: Use `await`

File: app/src/renderer/css/main.css [hunk 15]
```diff
@@ -239,6 +239,10 @@ body.is-tray-arrow-hidden {
 
 
 /* Aspect Ratio Controls */
```
Comment (sindresorhus) [inline app/src/renderer/css/main.css]: `Aspect ratio controls`

**Comment by sindresorhus** [comment]
Generally looks very good. I just commented about some nitpick :)

File: app/src/renderer/js/main.js (hunk 13 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: I got that from [here](https://github.com/wulkano/kap/blob/master/app/src/main/index.js#L32). I'm not 100% sure what its purpose is. I'm guessing to make up for the 1px of the border of the cropper window. Is that correct?

File: app/src/main/index.js (hunk 14 above)
Comment (karaggeorge) [reply app/src/main/index.js]: ea47eb8

File: app/src/renderer/js/main.js (hunk 5 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: ea47eb8

File: app/src/renderer/js/main.js (hunk 4 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: ea47eb8

File: app/src/renderer/js/main.js (hunk 12 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 45c39dd

File: app/src/renderer/js/main.js (hunk 10 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 451f413

File: app/src/renderer/js/main.js (hunk 1 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 451f413

File: app/src/renderer/views/main.pug (hunk 8 above)
Comment (karaggeorge) [reply app/src/renderer/views/main.pug]: 451f413

File: app/src/renderer/js/main.js (hunk 9 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 451f413

File: app/src/renderer/js/main.js (hunk 3 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 451f413

File: app/src/renderer/css/main.css (hunk 15 above)
Comment (karaggeorge) [reply app/src/renderer/css/main.css]: 451f413

File: app/src/renderer/js/main.js (hunk 6 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 5332f59

File: app/src/renderer/js/main.js (hunk 2 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 5332f59

File: app/src/renderer/js/main.js (hunk 7 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 1ce2c08

File: app/src/renderer/js/main.js (hunk 11 above)
Comment (karaggeorge) [reply app/src/renderer/js/main.js]: 211e9ca

**Comment by karaggeorge** [comment]
@sindresorhus 
Good catches! I fixed everything except for that `cropperWindowBuffer` variable, whose use I'm not 100% sure about.

File: app/src/renderer/js/main.js (hunk 13 above)
Comment (sindresorhus) [reply app/src/renderer/js/main.js]: It was originally added in https://github.com/wulkano/kap/commit/13a0cd6d66666cd2f5f395ebf001a1de4ced50d1

Comment (karaggeorge) [reply app/src/renderer/js/main.js]: ed705f7

**Comment by sallar** [comment]
@karaggeorge The `mac-windows` package throws an error. Both in this PR and also when installed standalone on the machine:

```
{ Error: Command failed: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true
dyld: Symbol not found: __TMSS
  Referenced from: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
  Expected in: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/libswiftCore.dylib
 in /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows

    at ChildProcess.exithandler (child_process.js:272:12)
    at ChildProcess.emit (events.js:159:13)
    at maybeClose (internal/child_process.js:943:16)
    at Socket.stream.socket.on (internal/child_process.js:363:11)
    at Socket.emit (events.js:159:13)
    at Pipe._handle.close [as _onclose] (net.js:568:12)
  killed: false,
  code: null,
  signal: 'SIGABRT',
  cmd: '/Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true' }
(node:10153) UnhandledPromiseRejectionWarning: TypeError: windows.filter is not a function
    at getWindows.then.then.windows (/Users/sallar/tmp/node_modules/mac-windows/index.js:41:22)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:160:7)
(node:10153) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:10153) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by karaggeorge** [comment]
@sallar Thanks for catching that. I installed it standalone and it worked on my machine with the published build, so that's a bit weird. (it obviously also worked in the app)

I updated the package (to make sure if an exception is thrown it's handled), re-built it and updated the dependency here. Can you check again?

**Comment by sallar** [comment]
@karaggeorge still no dice... running the binary throws this for me. I wonder if something on my machine is breaking it. but it shouldn't be because its a standalone binary right?

```
✔︎ tmp /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true
dyld: Symbol not found: __TMSS
  Referenced from: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
  Expected in: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/libswiftCore.dylib
 in /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
fish: '/Users/sallar/tmp/node_modules/…' terminated by signal SIGABRT (Abort)
```

**Comment by karaggeorge** [comment]
@sallar seems like other people have been having that issue:

>Oh, that's right. You can't (well, shouldn't) use binaries compiled with one version of Swift compiler in a runtime with another version.

My guess is that you have a different swift version than I do. I'll check to see what the best solution for this would be. Pretty new to swift, so I'm not quite sure.

**Comment by sallar** [comment]
@karaggeorge yes thats my guess too. I have swift 4. But what if users dont have swift or have another version of swift installed?

**Comment by albinekb** [comment]
@sallar the swift should be compiled before published to npm. Look at wulkano/aperture 👍

**Comment by karaggeorge** [comment]
@albinekb I did compile before I published it to npm, but I had Swift 3, so my compiled (with Swift 3) script couldn't run on a computer running Swift 4. 

@sallar, since this app only targets Mac users, I think it's a given a swift script will run on their computer (it's compiled into an executable). However, like you said what if they have a different version. I just upgraded to swift 4 and I'll try rebuilding now. I hope the Swift 4 compiled script will work for Swift 3. Otherwise maybe keeping both versions and checking the user's current environment?

**Comment by sallar** [comment]
@karaggeorge Thanks! Could you please do that rebuild and also fix this `main.js` conflict so we'll be able to merge?
@sindresorhus could you please check the PR again to see if we can approve for merging?

I need to do some work based on this PR and that can't be done until this is merged :)

Thanks all 🙌

**Comment by matheuss** [comment]
> so my compiled (with Swift 3) script couldn't run on a computer running Swift 4.

This doesn’t sound right - if you compile anything into a binary, you don’t shouldn’t need any of the build-time environment if you static link everything. I don’t remember having issues like that on aperture 🤔

**Comment by sallar** [comment]
Thats what I mean. Maybe this time release like this: `swift build -c release` maybe that helps. The build you’re currently doing is “debug” which might be the reason why it wont work.

**Comment by karaggeorge** [comment]
Like I said before, it's my first time working with Swift, so I'm don't really know how all of this works that well. I'll try building it as a release and let you know 👍

**Comment by sindresorhus** [comment]
@karaggeorge You need to build it with a statically bundled version of the Swift runtime as Swift is not yet ABI stable. Use this for building: `swift build --configuration=release -Xswiftc -static-stdlib`.

**Comment by karaggeorge** [comment]
@sindresorhus that did the trick 👍 

@sallar can you give it another try?

**Comment by sallar** [comment]
@karaggeorge Thanks! That did the trick and it now works. If you could also resolve that conflict so @sindresorhus can check again would be nice. Thanks!

**Comment by karaggeorge** [comment]
@sallar should be all good now!

File: app/src/renderer/js/main.js [hunk 16]
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+    }
+
+    if (appName === 'Fullscreen') {
+      let {width, height} = appData.get(appName);
+      // Need to get rid of the buffer because window can't be outside the screen limits
+      width -= app.kap.cropperWindowBuffer;
+      height -= app.kap.cropperWindowBuffer;
+      ipcRenderer.send('open-cropper-window', {width, height}, {x: 1, y: 1});
+    } else {
+      ipcRenderer.send('activate-application', appName, appData.get(appName));
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: `application` => `app`

File: app/src/renderer/js/main.js [hunk 17]
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  function clearApp() {
+    disabledAppOption.selected = true;
+  }
+
+  appSelector.append(createOption('Fullscreen'));
+  appSelector.append(createOption('──────────', {disabled: true}));
+  appSelector.addEventListener('change', handleAppChange);
+
+  async function loadApps() {
+    // Remove existing applications
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: apps

File: app/src/renderer/js/main.js [hunk 18]
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      if (option.value !== 'Fullscreen') {
+        option.remove();
+        appData.delete(option.value);
+      }
+    }
+
+    disabledAppOption.text = 'Loading…';
+    disabledAppOption.selected = true;
+
+    // Load applications
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: apps

File: app/package.json [hunk 19]
```diff
@@ -33,6 +33,7 @@
     "insight": "^0.8.3",
     "jquery": "^3.2.1",
     "lodash": "^4.17.4",
+    "mac-windows": "^0.5.4",
```
Comment (sindresorhus) [inline app/package.json]: You need to install it with Yarn, so it's added to the lockfile. Just run `yarn`.

**Comment by sindresorhus** [comment]
@karaggeorge This looks very good. Just some nitpick feedback left:

- When clicking the Kap menubar icon it shows "Select…" in the dropdown, then quickly for some milliseconds shows "Loading…", then back to "Select…" again.
- Would be nice to include icons in the list of app windows.
- When "Fullscreen" is selected in the dropdown, you can see that the text is cropped off too early, long before the down arrow icon. 
screenshot 2018-01-04 at 00 41 39
- When choosing "Fullscreen", the recording misses a few pixels on the edges of the screen.
screen shot 2018-01-04 at 00 46 46

The left is my screen, the right is the recording. I intentionally moved the window some pixel from the edge to show that the white stripe is not in the recording.

**Comment by sallar** [comment]
@sindresorhus @karaggeorge Im working on a PR that adds those icons and some other improvements.

**Comment by sindresorhus** [comment]
Ok, good to know. Ignore that here then :)

**Comment by sindresorhus** [comment]
Merging to unblock @sallar.

**Comment by sindresorhus** [comment]
Really nice work @karaggeorge 🙌

**Pull request merged**